### ISSUE TITLE
ClassCastException while searching using Lucene

### ISSUE THREAD

**Issue by vantonov1** [orientechnologies/orientdb#4257]
orient 2.0.10

If lucene index (OIndexTxAwareMultiValue with lucene as delegate) has changes inside transaction, seacrh throws exception

	at com.orientechnologies.orient.core.index.OCompositeKey.compareTo(OCompositeKey.java:37)
	at com.orientechnologies.common.comparator.ODefaultComparator.compare(ODefaultComparator.java:48)
	at java.util.TreeMap.getEntryUsingComparator(TreeMap.java:369)
	at java.util.TreeMap.getEntry(TreeMap.java:340)
	at java.util.TreeMap.get(TreeMap.java:273)
	at com.orientechnologies.orient.core.tx.OTransactionIndexChanges.getChangesPerKey(OTransactionIndexChanges.java:51)
	at com.orientechnologies.orient.core.index.OIndexTxAwareMultiValue.calculateTxIndexEntry(OIndexTxAwareMultiValue.java:446)
	at com.orientechnologies.orient.core.index.OIndexTxAwareMultiValue.get(OIndexTxAwareMultiValue.java:285)
	at com.orientechnologies.orient.core.index.OIndexTxAwareMultiValue.get(OIndexTxAwareMultiValue.java:40)
	at com.orientechnologies.lucene.operator.OLuceneTextOperator.executeIndexQuery(OLuceneTextOperator.java:47)
	at com.orientechnologies.orient.core.sql.OCommandExecutorSQLSelect.searchForIndexes(OCommandExecutorSQLSelect.java:1527)

Reason is simple - OCompositeKey implements Comparable &lt; OCompositeKey &gt;, but indexChanges holds String as key

**Comment by vantonov1** [comment]
still exists in 2.1-rc4

**Comment by nagarajasr** [comment]
@vantonov1 Could you provide more details how to reproduce the error?

**Comment by vantonov1** [comment]
final ODatabaseDocumentTx db = new ODatabaseDocumentTx("memory:default").open("admin", "admin");
        final OrientVertexType c1 = new OrientGraphNoTx(db).createVertexType("C1");
        c1.createProperty("p1", OType.STRING);
        c1.createIndex("p1", "FULLTEXT", null, null, "LUCENE", new String[]{"p1"});

        final OrientGraph graph = new OrientGraph(db);
        graph.begin();
        final OrientVertex result = graph.addVertex("class:C1");
        result.setProperty("p1", "abc");

        final String query = "select from C1 where [p1] lucene \"abc\" limit 1";
        final List<ODocument> vertices = ODatabaseRecordThreadLocal.INSTANCE.get().command(new OSQLSynchQuery<ODocument>(query)).execute();
        assert vertices != null && vertices.size() == 1;

**Comment by maggiolo00** [comment]
hi @vantonov1 
i'm on it

**Comment by maggiolo00** [comment]
@vantonov1  you can resolve
by committing before querying.

```
 final ODatabaseDocumentTx db = new ODatabaseDocumentTx("memory:default").open("admin", "admin");
    final OrientVertexType c1 = new OrientGraphNoTx(db).createVertexType("C1");
    c1.createProperty("p1", OType.STRING);
    c1.createIndex("p1", "FULLTEXT", null, null, "LUCENE", new String[]{"p1"});

    final OrientGraph graph = new OrientGraph(db);
    graph.begin();
    final OrientVertex result = graph.addVertex("class:C1");
    result.setProperty("p1", "abc");
   graph.commit();
    final String query = "select from C1 where [p1] lucene \"abc\" limit 1";
    final List<ODocument> vertices = ODatabaseRecordThreadLocal.INSTANCE.get().command(new OSQLSynchQuery<ODocument>(query)).execute();
    assert vertices != null && vertices.size() == 1;
```

The problem is querying within the transaction that uses a transaction cache for the changes and it will not use the lucene index

**Comment by vantonov1** [comment]
Sorry, could you provide more information. 

OrientDB has special logic, post-handling query result to make it aligned with transactional cache. But with lucene indices it fails with class cast - definitely it works not as expected

IMHO you have to exclude that logic and claim that it is not possible to do lucene queries inside transactions (it means that OrientDB + Lucene from local client is not usable in real world applications, business logic could be complicated), or fix OTransactionIndexChanges to work with Lucene

BTW, it is similar problem - https://groups.google.com/forum/#!topic/orient-database/Dw0zE2GwQhA

**Comment by maggiolo00** [comment]
The ClassCastException of course is a bug. 

The OrientDB transaction logic and Lucene are slightly different. When a TX is started OrientDB keeps track of the changes and if an index is involved OrientDB will ask to OTransactionIndexChanges first for requested value, without involving the lucene index as the value is not yet written in it until is committed.
BTW There is an issue open to improve the TX management, 

https://github.com/orientechnologies/orientdb-lucene/issues/46

**Comment by maggiolo00** [comment]
@vantonov1  

the first implementation of Tx support in plocal with Tx changes i available in develop
Will be released in OrientDB 2.2

**Comment by vantonov1** [comment]
Great!

Is 2.2.0-beta2 or RC planned before release? With that bug fixed, I'll be able to test the whole thing again, but it is better to test tagged build, not snapshot...

**Comment by maggiolo00** [comment]
@vantonov1 
yes i think will be released.

**Comment by vantonov1** [comment]
fixed in 2.2.0-beta
